### PR TITLE
Support searching by cancer type

### PR DESCRIPTION
**Pull request by Eric-Jiang1**
for #3415 

Adding the logic for search by cancer type, would it be ok to loop through entirely and not have a isMatch for when a match is found inside the for loop searching a list of cancers under evidence. 

This is because nothing would be found if we search for Leukemia, since it's not the name of any subtypes. It is the name for the main type, which will have multiple matches.

### CODE REVIEW

File: core/src/main/java/org/mskcc/cbio/oncokb/apiModels/CancerMatch.java [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+import org.mskcc.cbio.oncokb.model.Gene;
+import org.mskcc.cbio.oncokb.model.LevelOfEvidence;
+import org.mskcc.cbio.oncokb.model.TumorType;
+
+import java.util.HashSet;
+import java.util.Objects;
+import java.util.Set;
+
+
+public class CancerMatch {
```
Comment (zhx828) [inline core/src/main/java/org/mskcc/cbio/oncokb/apiModels/CancerMatch.java:13]:
```suggestion
public class CancerTypeMatch {
```
Let's stick to naming `CancerType`

File: core/src/main/java/org/mskcc/cbio/oncokb/model/TypeaheadQueryType.java [hunk 2]
```diff
@@ -4,5 +4,5 @@
  * Created by Hongxin Zhang on 8/20/20.
  */
 public enum TypeaheadQueryType {
-    GENE, VARIANT, DRUG, TEXT
+    GENE, VARIANT, DRUG, CANCER, TEXT
```
Comment (zhx828) [inline core/src/main/java/org/mskcc/cbio/oncokb/model/TypeaheadQueryType.java:7]:
```suggestion
    GENE, VARIANT, DRUG, CANCER_TYPE, TEXT
```

File: web/src/main/java/org/mskcc/cbio/oncokb/api/pvt/PrivateSearchApiController.java [hunk 3]
```diff
@@ -182,6 +189,101 @@ public ResponseEntity<LinkedHashSet<org.oncokb.oncokb_transcript.client.Drug>> s
...
+        Map<String, CancerMatch> result = new HashMap<>();
+
+        query = query.toLowerCase();
+         System.out.println("This is the query: " + query);
+
+        // cancer type not found, return an empty list
+        if (ApplicationContextSingleton.getTumorTypeBo().getByName(query) == null){
+            System.out.println("Nothing found");
+            return Collections.emptyList();
+        }
```
Comment (zhx828) [inline web/src/main/java/org/mskcc/cbio/oncokb/api/pvt/PrivateSearchApiController.java:206]:
```suggestion
        TumorType matchedTumorType = ApplicationContextSingleton.getTumorTypeBo().getByName(query);
        if ( matchedTumorType == null){
            System.out.println("Nothing found");
            return Collections.emptyList();
        }
```

File: web/src/main/java/org/mskcc/cbio/oncokb/api/pvt/PrivateSearchApiController.java [hunk 4]
```diff
@@ -182,6 +189,101 @@ public ResponseEntity<LinkedHashSet<org.oncokb.oncokb_transcript.client.Drug>> s
...
+
+        // definitely a matching cancer type, now searching evidence for relevant ones
+
+         for (Evidence evidence : evidences) {
+//             boolean isMatch = false;
+             for (TumorType cancer : TumorTypeUtils.findEvidenceRelevantCancerTypes(evidence)){
```
[the remark this answers is not in the document]
Comment (zhx828) [reply web/src/main/java/org/mskcc/cbio/oncokb/api/pvt/PrivateSearchApiController.java:212]: Use the `matchedTumorType` from above.
`TumorTypeUtils.findEvidenceRelevantCancerTypes(evidence).includes(matchedTumorType)`

**Review by zhx828 — CHANGES_REQUESTED** [review]

File: core/src/main/java/org/mskcc/cbio/oncokb/apiModels/CancerTypeMatch.java [hunk 5]
```diff
@@ -0,0 +1,35 @@
+package org.mskcc.cbio.oncokb.apiModels;
+
+import org.mskcc.cbio.oncokb.model.TumorType;
+
+import java.util.Objects;
+
+
+public class CancerTypeMatch extends LevelsOfEvidence {
+    private TumorType cancer;
```
Comment (zhx828) [inline core/src/main/java/org/mskcc/cbio/oncokb/apiModels/CancerTypeMatch.java:9]:
```suggestion
    private TumorType cancerType;
```

File: core/src/main/java/org/mskcc/cbio/oncokb/apiModels/LevelsOfEvidence.java [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+import org.mskcc.cbio.oncokb.model.Drug;
+import org.mskcc.cbio.oncokb.model.Gene;
+import org.mskcc.cbio.oncokb.model.LevelOfEvidence;
+import org.mskcc.cbio.oncokb.model.TumorType;
+
+import java.util.HashSet;
+import java.util.Objects;
+import java.util.Set;
+
+public class LevelsOfEvidence {
```
Comment (zhx828) [inline core/src/main/java/org/mskcc/cbio/oncokb/apiModels/LevelsOfEvidence.java:13]:
```suggestion
public class LevelsOfEvidenceMatch {
```

File: web/src/main/java/org/mskcc/cbio/oncokb/api/pvt/PrivateSearchApiController.java [hunk 7]
```diff
@@ -182,6 +190,108 @@ public ResponseEntity<LinkedHashSet<org.oncokb.oncokb_transcript.client.Drug>> s
...
+         }
+
+        // cancer type not found, return an empty list
+         if (matchedTumorTypes.isEmpty()){
+             System.out.println("Nothing found");
+             return Collections.emptyList();
+         }
+
+         // definitely a matching cancer type, now searching evidence for relevant ones
+         for (TumorType currMatchedCancer : matchedTumorTypes){
```
Comment (zhx828) [inline web/src/main/java/org/mskcc/cbio/oncokb/api/pvt/PrivateSearchApiController.java:234]: If no evidence is associated, we should still return the cancer type as a match. But just say in the frontend that no actionable gene is associated with this cancer type.

File: web/src/main/java/org/mskcc/cbio/oncokb/api/pvt/PrivateSearchApiController.java [hunk 8]
```diff
@@ -504,14 +616,31 @@ public int compare(TypeaheadSearchResp e1, TypeaheadSearchResp e2) {
...
+            result = LevelUtils.compareLevel(o1.getLevelOfEvidence(), o2.getLevelOfEvidence());
+            if(result == 0) {
+                result = o1.getGene().getHugoSymbol().compareTo(o2.getGene().getHugoSymbol());
+            }
+        }
+        return result;
+    }
+}
+
+class CancerMatchComp implements Comparator<CancerTypeMatch> {
```
Comment (zhx828) [inline web/src/main/java/org/mskcc/cbio/oncokb/api/pvt/PrivateSearchApiController.java:633]:
```suggestion
class CancerTypeMatchComp implements Comparator<CancerTypeMatch> {
```

File: core/src/main/java/org/mskcc/cbio/oncokb/apiModels/CancerTypeMatch.java [hunk 9]
```diff
@@ -0,0 +1,35 @@
+package org.mskcc.cbio.oncokb.apiModels;
+
+import org.mskcc.cbio.oncokb.model.TumorType;
+
+import java.util.Objects;
+
+
+public class CancerTypeMatch extends LevelsOfEvidence {
```
Comment (zhx828) [inline core/src/main/java/org/mskcc/cbio/oncokb/apiModels/CancerTypeMatch.java:8]: Nice

File: core/src/main/java/org/mskcc/cbio/oncokb/model/TypeaheadSearchResp.java [hunk 10]
```diff
@@ -11,6 +11,7 @@ public class TypeaheadSearchResp {
     private Set<Alteration> variants;
     private Set<TumorType> tumorTypes;
     private Drug drug;
+    private TumorType cancer;
```
Comment (zhx828) [inline core/src/main/java/org/mskcc/cbio/oncokb/model/TypeaheadSearchResp.java:14]:
```suggestion
    private TumorType cancerType;
```

File: web/src/main/java/org/mskcc/cbio/oncokb/api/pvt/PrivateSearchApiController.java [hunk 11]
```diff
@@ -182,6 +190,108 @@ public ResponseEntity<LinkedHashSet<org.oncokb.oncokb_transcript.client.Drug>> s
...
     }
 
+     private List<TypeaheadSearchResp> findMatchingCancerTypes(String query, Boolean exactMatch) {
+        // for now only finding exact matches
+         // got all evidences and level
+         System.out.println("This got to cancer type search ");
+         Set<Evidence> evidences = EvidenceUtils.getEvidenceByEvidenceTypesAndLevels(EvidenceTypeUtils.getTreatmentEvidenceTypes(), LevelUtils.getPublicLevels());
+         Map<String, CancerTypeMatch> result = new HashMap<>();
+
+         query = query.toLowerCase();
```
Comment (zhx828) [inline web/src/main/java/org/mskcc/cbio/oncokb/api/pvt/PrivateSearchApiController.java:200]: Please do a check on query if it's null. We also need to trim it.

File: web/src/main/java/org/mskcc/cbio/oncokb/api/pvt/PrivateSearchApiController.java [hunk 12]
```diff
@@ -182,6 +190,108 @@ public ResponseEntity<LinkedHashSet<org.oncokb.oncokb_transcript.client.Drug>> s
...
+         System.out.println("This is the query: " + query);
+
+         ArrayList<TumorType> matchedTumorTypes = new ArrayList<>();
+
+         if (exactMatch){
+             matchedTumorTypes.add(ApplicationContextSingleton.getTumorTypeBo().getByName(query));
+         }
+
+         if (matchedTumorTypes.isEmpty() && !exactMatch) {
```
Comment (zhx828) [inline web/src/main/java/org/mskcc/cbio/oncokb/api/pvt/PrivateSearchApiController.java:209]: It's essentially an else clause. 

```suggestion
         } else {
```

File: web/src/main/java/org/mskcc/cbio/oncokb/api/pvt/PrivateSearchApiController.java [hunk 13]
```diff
@@ -182,6 +190,108 @@ public ResponseEntity<LinkedHashSet<org.oncokb.oncokb_transcript.client.Drug>> s
...
+             for (Map.Entry<String, TumorType> currSubType : CacheUtils.getLowercaseSubtypeTumorTypeMap().entrySet()) {
```
Comment (zhx828) [inline web/src/main/java/org/mskcc/cbio/oncokb/api/pvt/PrivateSearchApiController.java:210]:
```suggestion
             for (Map.Entry<String, TumorType> currSubtype : CacheUtils.getLowercaseSubtypeTumorTypeMap().entrySet()) {
```

**Review by zhx828 — CHANGES_REQUESTED** [review]
- Just a reminder, please remove all necessary logs before merging.
- Make sure to keep the consistency when naming variables

File: web/src/main/java/org/mskcc/cbio/oncokb/api/pvt/PrivateSearchApiController.java [hunk 14]
```diff
@@ -182,6 +190,108 @@ public ResponseEntity<LinkedHashSet<org.oncokb.oncokb_transcript.client.Drug>> s
...
+        typeaheadSearchResp.setCancer(cancerMatch.getCancer());
+
+        if (LevelUtils.isSensitiveLevel(cancerMatch.getLevelOfEvidence())) {
+            typeaheadSearchResp.setHighestSensitiveLevel(cancerMatch.getLevelOfEvidence().getLevel());
+        } else {
+            typeaheadSearchResp.setHighestResistanceLevel(cancerMatch.getLevelOfEvidence().getLevel());
+        }
+        typeaheadSearchResp.setQueryType(TypeaheadQueryType.CANCER_TYPE);
+
+        typeaheadSearchResp.setLink("actionableGenes#hugoSymbol=" + cancerMatch.getGene().getHugoSymbol() + "&sections=FDA&tumorType=" + cancerMatch.getCancer().getCode());
```
Comment (zhx828) [inline web/src/main/java/org/mskcc/cbio/oncokb/api/pvt/PrivateSearchApiController.java:273]:
```suggestion
        typeaheadSearchResp.setLink("actionableGenes#hugoSymbol=" + cancerMatch.getGene().getHugoSymbol() + "&sections=Tx&tumorType=" + cancerMatch.getCancer().getCode());
```

File: web/src/main/java/org/mskcc/cbio/oncokb/api/pvt/PrivateSearchApiController.java (hunk 11 above)
Comment (Eric-Jiang1) [reply web/src/main/java/org/mskcc/cbio/oncokb/api/pvt/PrivateSearchApiController.java:200]: we have a check for this in searchTypeAheadGet before the query gets passed to any of the other search functions

File: web/src/main/java/org/mskcc/cbio/oncokb/api/pvt/PrivateSearchApiController.java (hunk 14 above)
Comment (zhx828) [reply web/src/main/java/org/mskcc/cbio/oncokb/api/pvt/PrivateSearchApiController.java]: This hasn't been updated.

**Comment by zhx828** [comment]
Few issues I noticed based on the latest code
- the search result only shows Level 1 data, which it should show all.
- When search keyword, I would hope to see the matches start with the keyword to be ranked higher than others. For instance, if I search `Mel`, I would want to see Melanoma on the top.

**Review by zhx828 — CHANGES_REQUESTED** [review]
Could you do a squash of your commits and rebase to the rc branch? Let me know if you don't know how to do it.

**Pull request merged**